### PR TITLE
feat(order-sync): add line item state support to order synchronization

### PR DESCRIPTION
**Pull request by Siilwyn**
- [x] commit messages are commitizen-friendly
- [x] code is unit tested
- [x] code is integration tested
- [x] public code is documented
- [x] code is reviewed by at least one person
- [x] code satisfies linter rules

### CODE REVIEW

File: src/coffee/sync/utils/order.coffee [hunk 1]
```diff
@@ -117,6 +117,37 @@ class OrderUtils extends BaseUtils
...
+    
+    data = new_obj || old_obj
+    
+    # Note the lineItem.state contains 'from/to' instead of just the data
+
+    if data.lineItems
+      data.lineItems.forEach (lineItem) ->
+        if (lineItem.state)
+          lineItem.state.forEach (state) ->
+            if (state.fromState && state.toState)
```
Comment (hisabimbola) [inline src/coffee/sync/utils/order.coffee]: Since this is an object and empty object is truthy you might want to check the size or values in the object.

Comment (Siilwyn) [inline src/coffee/sync/utils/order.coffee]: The API will return an error if the content of one of the fields is incorrect.

File: src/coffee/sync/utils/order.coffee [hunk 2]
```diff
@@ -117,6 +117,37 @@ class OrderUtils extends BaseUtils
...
+              actions.push(
+                {
+                  action: "transitionLineItemState",
+                  lineItemId: lineItem.id,
+                  quantity: state.quantity,
+                  fromState: state.fromState,
+                  toState: state.toState,
+                  actualTransitionDate: state.actualTransitionDate
```
Comment (Siilwyn) [inline src/coffee/sync/utils/order.coffee]: Sharp! Will check it.

**Comment by Siilwyn** [comment]
I do not understand how the coverage decreased, it looks like coveralls says [this line](https://github.com/sphereio/sphere-node-sdk/blob/support-order-line-items/src/coffee/sync/utils/order.coffee#L147) is not covered but as far as I can see it is being check both resolving to false an true in [these unit tests](https://github.com/sphereio/sphere-node-sdk/blob/support-order-line-items/src/spec/sync/utils/order.spec.coffee#L416-L492).
:"|

**Comment by LEQADA** [comment]
[On this line](https://github.com/sphereio/sphere-node-sdk/blob/support-order-line-items/src/spec/sync/utils/order.spec.coffee#L430)  you put extra comma. I bet that you wanted to add `actualTransitionDate: 'someDateHere'` but forgot :)
That's why it's not cover [this line](https://github.com/sphereio/sphere-node-sdk/blob/support-order-line-items/src/coffee/sync/utils/order.coffee#L147)
And if you'll add that `actualTransitionDate` to `expectedActions` then I think coveralls will stop screaming

**Comment by Siilwyn** [comment]
@LEQADA thank you for the suggestion. That comma is there just for trailing, a habit I got into since our new modules require is for the linter. As for adding it that is what [this test](https://github.com/sphereio/sphere-node-sdk/blob/support-order-line-items/src/spec/sync/utils/order.spec.coffee#L468-L488) is for right? Or am I missing something?

File: src/spec/integration/orders-sync.spec.coffee [hunk 3]
```diff
@@ -212,6 +222,70 @@ describe 'Integration Orders Sync', ->
       done()
     .catch (error) -> done(_.prettify(error))
 
+  it 'should sync line items', (done) ->
+    orderNew = _.deepClone @order
+    
+    orderNew.lineItems[0].state = [
+      {
+        quantity: 1,
+        fromState: {
```
Comment (PhilippSpo) [inline src/spec/integration/orders-sync.spec.coffee]: Why are you doing this? This means that when I use the SDK to generate update actions for line item states, I have to provide the `fromState` and `toState`. This seems wrong. We can just look ad the "old" order to get the from state and at the "new" order to get the new state, right?

**Comment by Siilwyn** [comment]
After talking with Philipp we have decided to implement this in the module that is updating because this is not really diff'ing the data and thus breaks with the concept of the sync modules in the SDK.

**Pull request closed without merging**